File: 프로그래머스/3/150366. 표 병합/표 병합.py
# from collections import defaultdict

def solution(commands):
    answer = []
    # print(commands)
    board,parents=init()
    
    for command in commands:
        com=command.split(" ")
        
        if com[0]=="UPDATE":
            if len(com)==4:
                update1(int(com[1]),int(com[2]),com[3],board,parents)
        if com[0]=="UPDATE":
            if len(com)==3:
                update2(com[1],com[2],board)
        if com[0]=="MERGE":
            merge(int(com[1]),int(com[2]),int(com[3]),int(com[4]),board,parents)
        if com[0]=="UNMERGE":
            unmerge(int(com[1]),int(com[2]),board,parents)
        if com[0]=="PRINT":
            answer.append(print_board(int(com[1]),int(com[2]),board,parents))
    # print(parents)
    return answer
def init():
    board=[[""for _ in range(51)] for _ in range(51)]
    parents=[[[]for _ in range(51)] for _ in range(51)]
    for r in range(51):
        for c in range(51):
            parents[r][c].append([r,c])
            parents[r][c].append([])
            parents[r][c][1].append([r,c])
            
    return board,parents

def update1(r,c,value,board,parents):
    R,C=find_parents([r,c],parents)
    board[R][C]=value
def update2(value1,value2,board):
    for r in range(51):
        for c in range(51):
            if board[r][c]==value1:
                board[r][c]=value2
                
def merge(r1,c1,r2,c2,board,parents):
    R1,C1=find_parents([r1,c1],parents)
    R2,C2=find_parents([r2,c2],parents)
    if (R1,C1)==(R2,C2):
        return
    if board[R1][C1]:
        parents[R2][C2][0]=[R1,C1]
        parents[R1][C1][1]+=parents[R2][C2][1]
        return 
    
    if board[R2][C2]:
        parents[R1][C1][0]=[R2,C2]
        parents[R2][C2][1]+=parents[R1][C1][1]
        return
    else:
        parents[R1][C1][0]=[R2,C2]
        parents[R2][C2][1]+=parents[R1][C1][1]
        return
        
    
def find_parents(node,parents):
    r,c=node
    if node==parents[r][c][0]:
        return parents[r][c][0]
    parents[r][c][0]=find_parents(parents[r][c][0],parents)
    return parents[r][c][0]


def print_board(r,c,board,parents):
    R,C=find_parents([r,c],parents)
    if board[R][C]:
        return board[R][C]
    else:
        return "EMPTY"
def unmerge(r,c,board,parents):
    # for p in parents:
    #     print(p)
    # print(board)
    R,C=find_parents([r,c],parents)
    value=board[R][C]
    for child_r,child_c in parents[R][C][1]:
        parents[child_r][child_c][0]=[child_r,child_c]
        parents[child_r][child_c][1]=[[child_r,child_c]]
        board[child_r][child_c]=""
    # print("!!")
    # for p in parents:
    #     print(p)
    board[r][c]=value
    # print(board)
    
    

File: 프로그래머스/3/150366. 표 병합/표 병합.java
import java.util.*;
class Solution {
    String[][] board=new String[51][51];
    Node[][] parents=new Node[51][51];
    ArrayList<String> ans=new ArrayList<>();
    public String[] solution(String[] commands) {
        for(int r=1;r<=50;r++){
            for(int c=1;c<=50;c++){
                parents[r][c]=new Node(r,c);
            }
        }
        for(int i=0;i<commands.length;i++){
            String[] tmp=commands[i].split(" ");
            String op=tmp[0];
            if(op.equals("UPDATE")){
                // System.out.println(Arrays.toString(tmp));
                if(tmp.length==4){
                    int r=Integer.parseInt(tmp[1]);
                    int c=Integer.parseInt(tmp[2]);
                    String v=tmp[3];
                    update(r,c,v);
                }
                else{
                    update(tmp[1],tmp[2]);
                }
            }
            if(op.equals("MERGE")){
                // System.out.println(Arrays.toString(tmp));
                int r1=Integer.parseInt(tmp[1]);
                int c1=Integer.parseInt(tmp[2]);
                int r2=Integer.parseInt(tmp[3]);
                int c2=Integer.parseInt(tmp[4]);
                merge(r1,c1,r2,c2);
            }
            if(op.equals("UNMERGE")){
                // System.out.println(Arrays.toString(tmp));
                int r1=Integer.parseInt(tmp[1]);
                int c1=Integer.parseInt(tmp[2]);
                unmerge(r1,c1);
            }
            if(op.equals("PRINT")){
                // System.out.println(Arrays.toString(tmp));
                int r1=Integer.parseInt(tmp[1]);
                int c1=Integer.parseInt(tmp[2]);
                ans.add(print(r1,c1));
            }
            // for(int n=1;n<=50;n++){
            //     System.out.println(Arrays.toString(parents[n]));
            // }
        }
        String[] answer = new String[ans.size()];
        for(int i=0;i<ans.size();i++){
            answer[i]=ans.get(i);
        }
        return answer;
    }
    Node findParent(int r,int c) {
        if(parents[r][c]!=parents[r][c].parent){
            return parents[r][c].parent=findParent(parents[r][c].parent.r,parents[r][c].parent.c);
        }
        return parents[r][c];
    }
    void update(int r,int c,String value){
        Node parent=findParent(r,c);
        parent.value=value;
    }
    void update(String value1,String value2){
        for(int i=1;i<=50;i++){
            for(int j=1;j<=50;j++){
                if(findParent(i,j).value!=null && findParent(i,j).value.equals(value1)){
                    findParent(i,j).value=value2;
                }
            }
        }
    }
    
    void merge(int r1,int c1,int r2,int c2){
        Node parent1=findParent(r1,c1);
        Node parent2=findParent(r2,c2);
        
        if(parent1.value!=null && parent2.value!=null){
            parent2.parent=parent1;
            return;
        }
        if(parent1.value!=null){
            parent2.parent=parent1;
            return;
        }
        if(parent2.value!=null){
            parent1.parent=parent2;
            return;
        }
        if(parent1.value==null && parent2.value==null){
            parent2.parent=parent1;
            return;
        }
    }
    String print(int r,int c){
        Node parent=findParent(r,c);
        // System.out.println(parent.value);
        if(parent.value!=null) return parent.value;
        return "EMPTY";
    }
    void unmerge(int r,int c){
        Node parent=findParent(r,c);
        
        String tmp=parent.value;
        for(int i=1;i<=50;i++){
            for(int j=1;j<=50;j++){
                findParent(i,j);
            }
        }
        for(int i=1;i<=50;i++){
            for(int j=1;j<=50;j++){
                // System.out.println("ㅡㅡㅡㅡㅡㅡㅡ");
                // System.out.printf("%d %d\n",i,j);
                // System.out.println(findParent(i,j));
                // System.out.println(parents[i][j].parent);
                if(findParent(i,j)==parent){
                    // System.out.printf("%d %d\n",parents[i][j].r,parents[i][j].c);
                    parents[i][j].parent=parents[i][j];
                    parents[i][j].value=null;
                    // System.out.println(findParent(i,j));
                }
                
            }
        }
        parents[r][c].value=tmp;
        
    }
    class Node{
        int r;
        int c;
        String value;
        Node parent;
        Node(int r,int c){
            this.r=r;
            this.c=c;
            parent=this;
        }
        public String toString(){
            return this.r+" "+this.c;
        }
    }
}